### DISCUSSION TITLE
Proper way to automatically start fluidsynth and jackd (Raspberry Pi OS)

### DISCUSSION THREAD

**Discussion by mojca** [FluidSynth/fluidsynth#1297] [Q&A] [answered]
I have a simple arduino-based MIDI instrument connected to a Raspberry PI, running 64-bit Raspberry Pi OS based on Debian 12.
RPi is also connected to a PCM5120 and I followed [these instructions](https://blog.himbeer.me/2018/12/27/how-to-connect-a-pcm5102-i2s-dac-to-your-raspberry-pi/) to set it up.

I can launch two `ssh` shells to the RPi and run:
```sh
/usr/bin/jackd --timeout 2000 --realtime --realtime-priority 95 -d alsa --device hw:sndrpihifiberry --period 64 --nperiods 2 --rate 44100
```
followed by
```
fluidsynth --audio-driver=jack --midi-driver=alsa_seq --sample-rate=44100 --connect-jack-outputs --audio-channels=2 -is /usr/share/sounds/sf2/my_own_sound_font.sf2
```
in the second shell. This way I can get very low latency when playing the midi instrument.

However I don't seem to be able to figure out how to automatically start any of the two (`jackd` or `fluidsynth`) via `systemd` or via some other equivalent mechanism that would let me start the MIDI synthesis automatically.

I first managed to get the low latency working on [Patchbox OS](https://blokas.io/patchbox-os/) by running Qsynth, but I would like to get it working on barebone minimal Debian running in headless mode, without any UI.

I saw that Patchbox has something similar to the following contents in  `/etc/jackdrc` (below I wrote the options that I though I needed myself):
```
$ cat
#!/bin/sh
# exec is used to 'morph' the shell interpreter process into jackd process, saving some system resources.
exec /usr/bin/jackd -v --timeout 2000 --realtime --realtime-priority 95 -d alsa --device hw:sndrpihifiberry --period 64 --nperiods 2 --rate 44100
```

If I don't start `jackd` upfront, `fluidsynth` does seem to attempt starting `jackd`, but something weird starts going on. In particular I don't quite understand `exec of JACK server (command = "#!/bin/sh") failed: No such file or directory`. And `Cannot connect to server socket err = No such file or directory` also doesn't help me much.

```
$ fluidsynth --audio-driver=jack --midi-driver=alsa_seq --sample-rate=44100 --connect-jack-outputs --audio-channels=2 -is /usr/share/sounds/sf2/my_own_sound_font.sf2
FluidSynth runtime version 2.3.1
Copyright (C) 2000-2022 Peter Hanappe and others.
Distributed under the LGPL license.
SoundFont(R) is a registered trademark of Creative Technology Ltd.

fluidsynth: warning: No preset found on channel 9 [bank=128 prog=0]
fluidsynth: Connection of Midi Through Port-0 succeeded
fluidsynth: Connection of Arduino NANO 33 IoT MIDI 1 succeeded
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
exec of JACK server (command = "#!/bin/sh") failed: No such file or directory
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
fluidsynth: error: Failed to connect to Jack server.
Failed to create the audio driver. Giving up.
```

I tried starting `jackd` with a service like the one I found [here](https://github.com/soulik/systemd-config/blob/master/user/jack.service), but there are two problems:
* `/usr/bin/jack_control start` completely ignores the parameters from `/etc/jackdrc` (I assume I need to configure `jackd` elsewhere?)
* `/usr/bin/jack_control` refuses to work when run as `root`, but adding `User=myusername` under `[Service]` doesn't seem to help either

Debian comes with a ready-made systemd file for `fluidsynth`. I can edit `/etc/default/fluidsynth` to pass the correct parameters, but I don't come much further until I'm able to figure out how to get the `jackd` working properly/automatically.
```
sudo systemctl enable /usr/lib/systemd/user/fluidsynth.service
sudo systemctl start fluidsynth.service
```

**Comment by derselbst** [comment, marked as answer]
Fluidsynth doesn't actively starts up jack. Merely Jack starts itself (I think there's an option to disable that, but fluidsynth hasn't set it). I don't know about the underlying magic and related Jack commandline logs you're seeing, sry.

Regarding fluidsynth, pls. verify that your `/usr/lib/systemd/user/fluidsynth.service` file is same (or similar) to ours:
https://github.com/FluidSynth/fluidsynth/blob/bd04f3f6c81a114f3ad52c8e29be3b44dccd010f/fluidsynth.service.in

The line that starts with `ExecStart` is crucial. You'll notice a few environment variables in that line, that you can populate via EnvironmentFiles with our favorite fluidsynth options, like so:
https://github.com/FluidSynth/fluidsynth/blob/bd04f3f6c81a114f3ad52c8e29be3b44dccd010f/fluidsynth.conf.in

Usually, a system-wide and user-local EnvironmentFile exists. The exact location is also written in the .service file. It's up to you which to choose.

When done configuring, you should always start fluidsynth as a user-service:

`systemctl enable --user fluidsynth.service`
`systemctl start --user fluidsynth.service`

I would expect a similar setup procedure for the Jack Daemon (i.e. user service, EnvironmentFiles, etc.). But I have no idea how to configure jack + systemd properly.

**Reply by mojca** [reply]
Thank you very much for pointing me to the `--user` option. This is the one that I was in fact missing. Though I'm still unable to get everything working together.

I copied the following `jackd` service file from somewhere. I suspect it's not entirely correct, but at least I was now able to start it manually (after logging in via `ssh`) with the `--user` flag.
```
$ cat /usr/lib/systemd/user/jackd.service
[Unit]
Description=JACK 2
Before=sound.target
Before=pulseaudio.service
Requires=dbus.socket
After=usb-reset.service

[Service]
Type=dbus
BusName=org.jackaudio.Controller
ExecStart=/usr/bin/jack_control start
ExecStop=/usr/bin/jack_control stop
ExecStopPost=/usr/bin/pkill -9 jackdbus
SuccessExitStatus=0
RemainAfterExit=true
#Restart=always
#RestartSec=5

[Install]
WantedBy=sound.target
```

(I don't see anything corresponding to `BusName=org.jackaudio.Controller` inside `/usr/share/dbus-1/`)

```
$ systemctl --user status jackd.service
● jackd.service - JACK 2
     Loaded: loaded (/usr/lib/systemd/user/jackd.service; enabled; preset: enabled)
     Active: active (exited) since Wed 2024-01-03 19:01:22 CET; 9min ago
   Main PID: 1945 (code=exited, status=0/SUCCESS)
        CPU: 181ms

Jan 03 19:01:21 raspiccordion systemd[862]: Starting jackd.service - JACK 2...
Jan 03 19:01:22 raspiccordion jack_control[1945]: --- start
Jan 03 19:01:22 raspiccordion systemd[862]: Started jackd.service - JACK 2.
```

It does not work automatically after rebooting though, most likely because it's being started before DBus is ready.

```
$ systemctl --user status jackd.service
○ jackd.service - JACK 2
     Loaded: loaded (/usr/lib/systemd/user/jackd.service; enabled; preset: enabled)
     Active: inactive (dead)
$ journalctl --user -u jackd
Jan 03 21:12:12 raspiccordion systemd[862]: Starting jackd.service - JACK 2...
Jan 03 21:12:13 raspiccordion jack_control[2305]: --- start
Jan 03 21:12:13 raspiccordion systemd[862]: Started jackd.service - JACK 2.
Jan 03 21:16:19 raspiccordion systemd[862]: Stopping jackd.service - JACK 2...
Jan 03 21:16:19 raspiccordion jack_control[2707]: Traceback (most recent call last):
Jan 03 21:16:19 raspiccordion jack_control[2707]:   File "/usr/bin/jack_control", line 414, in <module>
Jan 03 21:16:19 raspiccordion jack_control[2707]:     sys.exit(main())
Jan 03 21:16:19 raspiccordion jack_control[2707]:              ^^^^^^
Jan 03 21:16:19 raspiccordion jack_control[2707]:   File "/usr/bin/jack_control", line 397, in main
Jan 03 21:16:19 raspiccordion jack_control[2707]:     bus = dbus.SessionBus()
Jan 03 21:16:19 raspiccordion jack_control[2707]:           ^^^^^^^^^^^^^^^^^
Jan 03 21:16:19 raspiccordion jack_control[2707]:   File "/usr/lib/python3/dist-packages/dbus/_dbus.py", line 209, in __new__
Jan 03 21:16:19 raspiccordion jack_control[2707]:     return Bus.__new__(cls, Bus.TYPE_SESSION, private=private,
Jan 03 21:16:19 raspiccordion jack_control[2707]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jan 03 21:16:19 raspiccordion jack_control[2707]:   File "/usr/lib/python3/dist-packages/dbus/_dbus.py", line 99, in __new__
Jan 03 21:16:19 raspiccordion jack_control[2707]:     bus = BusConnection.__new__(subclass, bus_type, mainloop=mainloop)
Jan 03 21:16:19 raspiccordion jack_control[2707]:           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jan 03 21:16:19 raspiccordion jack_control[2707]:   File "/usr/lib/python3/dist-packages/dbus/bus.py", line 120, in __new__
Jan 03 21:16:19 raspiccordion jack_control[2707]:     bus = cls._new_for_bus(address_or_type, mainloop=mainloop)
Jan 03 21:16:19 raspiccordion jack_control[2707]:           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jan 03 21:16:19 raspiccordion jack_control[2707]: dbus.exceptions.DBusException: org.freedesktop.DBus.Error.NoReply: Did not receive a reply. Possible causes include: the remote application did not send a >
Jan 03 21:16:19 raspiccordion systemd[862]: jackd.service: Control process exited, code=exited, status=1/FAILURE
Jan 03 21:16:19 raspiccordion systemd[862]: jackd.service: Control process exited, code=exited, status=1/FAILURE
Jan 03 21:16:19 raspiccordion systemd[862]: jackd.service: Failed with result 'exit-code'.
Jan 03 21:16:19 raspiccordion systemd[862]: Stopped jackd.service - JACK 2.
```

I tried to come up with something by following https://stackoverflow.com/questions/31702465/how-to-define-a-d-bus-activated-systemd-service, but I must have messed up completely.

If I start the above `jackd` service manually, then the combination with

```
$ cat ~/.config/fluidsynth
SOUND_FONT=/usr/share-sounds/sf2/my_own_sound_font.sf2
OTHER_OPTS='--audio-driver=jack --midi-driver=alsa_seq --sample-rate=44100 --gain=0.2 --connect-jack-outputs --audio-channels=2 -is'
```

and the service file that looks pretty much the same as your pointer seem to work fine.

(Initially I was getting `fluidsynth: error: Failed to bind server socket: 98`, but it turned out that I accidentally had another instance of `fluidsynth` started up automatically from `/etc/rc.local` as a "leftover from earlier attempts".)

So it seems that I'm only missing a way to properly start `jackd` now.

After you explicitly pointed me to the service file for `fluidsynth`, I noticed `pipewire.service` (I believe this is present out-of-the-box on Debian 12, but it wasn't in an earlier version of Debian). It looks like pipewire could be a replacement for `jackd`. I followed https://wiki.debian.org/PipeWire#Debian_12 and ran
```
sudo apt install wireplumber
systemctl --user --now enable wireplumber.service
```
but I would have to figure out how to get it working with fluidsynth as it complained about missing support for pipewire. I tried replacing `--audio-driver=jack` with `--audio-driver=pipewire`, but:
```
fluidsynth: error: Couldn't find the requested audio driver 'pipewire'.
```
The advantage would be the the service file for `wireplumber` already ships with the system (I trust it a lot more) and seems to be started correctly, while the `jackd` service file is most likely not written correctly, and the `jack_control` command also behaves in a somewhat weird way (it ignores the settings inside `/etc/jackdrc`, it is supposed to take care of starting `jackd` once, but it doesn't know in case it has already been started via some other way, etc.).

(On a second micro SD card we installed the fully blown OS. By configuring auto-login into the GUI and by starting `Qsynth` automatically after login, we do get the desired result with both `jackd` and `fluidsynth` being properly started by `Qsynth` in the background, without configuring any services whatsoever. But I don't particularly like all the extra overhead, and it also significantly prolongs the time needed between starting the device and being ready to play music.)

**Reply by derselbst** [reply]
I can't say anything about your jack and systemd problem, sry.

`fluidsynth: error: Couldn't find the requested audio driver 'pipewire'.`

That error indicates that fluidsynth was compiled without pipewire support (i.e. without libpipewire-dev being installed when compiling fluidsynth). You can either report this limitation to the package maintainer of your distro, or compile fluidsynth yourself locally.

**Reply by mojca** [reply]
I now found an excellent article explaining some details of DBus:
* https://www.baeldung.com/linux/systemd-session-dbus-headless-setup

By following this article I was finally able to get to a working system. I did multiple things, but probably the most relevant part was enabling auto-login:
```bash
loginctl enable-linger user
```
adding
```bash
if test -z "$DBUS_SESSION_BUS_ADDRESS"; then
    # Only start D-Bus if it's not already running
    eval `dbus-launch --sh-syntax --exit-with-session`
fi
```
to `.profile`, maybe adding some `[Install]` targets to existing `systemd` services, and finally here's the version of `/usr/lib/systemd/user/jackd.service` that I ended up using successfully (I was missing a bunch of `ExecStartPost` commands, since `jack_control` ignores `/etc/jackdrc`):
```config
[Unit]
Description=JACK 2
Before=sound.target
Requires=dbus.socket
After=usb-reset.service

[Service]
Type=dbus
BusName=org.jackaudio.Controller
ExecStart=/usr/bin/jack_control start
ExecStartPost=jack_control ds alsa
ExecStartPost=jack_control dps device hw:0,0
ExecStartPost=jack_control dps period 64
ExecStartPost=jack_control dps nperiods 2
ExecStartPost=jack_control dps rate 44100
ExecStop=/usr/bin/jack_control stop
ExecStopPost=/usr/bin/pkill -9 jackdbus
SuccessExitStatus=0
RemainAfterExit=true
Restart=on-failure
RestartSec=5

[Install]
WantedBy=sound.target
```

And in any case the `--user` flag was the crucial one to enable and run both services (thank you very much for that hint):
```
systemctl --user enable /usr/lib/systemd/user/jackd.service
systemctl --user enable /usr/lib/systemd/user/fluidsynth.service
systemctl --user start jackd
systemctl --user start fluidsynth
```
